enhancements/machine-config/mco-flattened-ignition.md

```diff
@@ -0,0 +1,176 @@
+so they may be consumed during deployment.
+
+The proposal is to create a third managed artifact, which is the flattened
+combination of the pointer and rendered ignition configs, so that when
+required it may be consumed instead of requiring runtime download of the rendered
+config via the pointer URL.
+
+Open Questions for discussion:
+  * Should the new artifact be another MachineConfig or a Secret or some other entity?
+  * How do we resolve the issue of CAPBM not being able to access the MCS,
+    can we reliably determine the MCO managed resource and reference it directly?
+  * Does this approach fit with discussions around declarative network
+    configuration (for day-1 and day-2)?
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+## Alternatives
+
+### Native ignition support for early-network config
+
+This was initially proposed via [an ignition issue](https://github.com/coreos/ignition/issues/979)
+where we had some good discussion, but ultimately IIUC the idea of adding a new
+interface for early network configuration to ignition was rejected.
+
+### Implement config flattening in platform-specific repos
+
+The ignition configuration is consumed at two points during deployment for IPI baremetal
+deployments:
+
+  * Via terraform (when the installer deploys the controlplane
+    nodes using [terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic).
+  * Workers are then deployed via the machine-api, using a
+    [baremetal Cluster API Provider](https://github.com/openshift/cluster-api-provider-baremetal/)
+
+The ironic terraform provider is designed to be generic, and thus we would prefer
+not to add OS specific handling of the user-data there, but it would be technically
+possible to do the flattening in some terraform provider, and duplicate that
+logic in the cluster-api-provider-baremetal.
+
+This option is probably suboptimal though, since it would require duplicate code,
+and knowledge of e.g ignition version/interfaces outside of the MCO managed
+configs.  It also would not be available for any other platforms or UPI usage.
+
+### Add support for injecting NM config to IPI
+
+This might be possible using a recently added [ironic python agent feature](https://specs.openstack.org/openstack/ironic-specs/specs/approved/in-band-deploy-steps.html)
+using that interface, it could be possible to inject network configuration in a
+similar way to coreos-install.
+
+However this is a very bleeding edge feature (not yet documented), and it means
+maintaining a custom interface that would be different to both the proven
+uncustomized OSP deploy ramdisk, and the coreos-deploy toolchain.
+
+### Convert IPI to coreos-install based deployment
+
+Long term, from an OpenShift perspective, this probably makes sense, however
```

Comment:
The way I'd say it is that there's massive overlap between the two - they're doing the same thing really.

Reply:
Yeah agreed, the issue is the well-tested path for Ironic at this point is to use the ironic community supported agent/ramdisk (which is why I said from an OpenShift perspective, we want to support a single solution for OpenShift, but the Ironic community will still want to maintain the existing solution).

This is solveable but is likely to take us some time (maybe create an FCOS based upstream CI job, deal with interface differences, etc) so it's being discounted as a viable solution in the 4.7 timeframe.

Reply:
I added a comment to clarify that there is overlap, and we're agreed that long-term this makes sense.